lib/src/runner/reporter/xunit.dart

```diff
@@ -0,0 +1,363 @@
+  /// when the reporter is paused.
+  var _stopwatchStarted = false;
+
+  /// Whether the reporter is paused.
+  var _paused = false;
+
+  /// The set of all subscriptions to various streams.
+  final _subscriptions = new Set<StreamSubscription>();
+
+  /// Record of all tests run.
+  Map _testCases = {};
+
+  /// Number of failures that occur during the test run.
+  int _failureCount = 0;
+
+  /// Number of errors that occur during the test run.
+  int _errorCount = 0;
+
+  /// Map containing Xunit hierarchy.
+  Map _groupStructure = {};
+
+  /// Watches the tests run by [engine] and records its results.
+  static XunitReporter watch(Engine engine, {bool verboseTrace: false}) {
+    return new XunitReporter._(engine, verboseTrace: verboseTrace);
+  }
+
+  XunitReporter._(this._engine, {bool verboseTrace: false}) {
+    _subscriptions.add(_engine.onTestStarted.listen(_onTestStarted));
+
+    /// Convert the future to a stream so that the subscription can be paused or
+    /// canceled.
+    _subscriptions.add(_engine.success.asStream().listen(_onDone));
+  }
+
+  void pause() {
+    if (_paused) return;
+    _paused = true;
+
+    _stopwatch.stop();
+
+    for (var subscription in _subscriptions) {
+      subscription.pause();
+    }
+  }
+
+  void resume() {
+    if (!_paused) return;
+    if (_stopwatchStarted) _stopwatch.start();
+
+    for (var subscription in _subscriptions) {
+      subscription.resume();
+    }
+  }
+
+  void cancel() {
+    for (var subscription in _subscriptions) {
+      subscription.cancel();
+    }
+    _subscriptions.clear();
+  }
+
+  /// A callback called when the engine begins running [liveTest].
+  void _onTestStarted(LiveTest liveTest) {
+    if (liveTest.suite is! LoadSuite) {
+      if (!_stopwatch.isRunning) _stopwatch.start();
+
+      // The engine surfaces load tests when there are no other tests running,
+      // but because the expanded reporter's output is always visible, we don't
+      // emit information about them unless they fail.
+      _subscriptions.add(liveTest.onStateChange
+          .listen((state) => _onStateChange(liveTest, state)));
+    } else if (_engine.active.length == 1 &&
+        _engine.active.first == liveTest &&
+        liveTest.test.name.startsWith("compiling ")) {}
+
+    _subscriptions.add(liveTest.onError
+        .listen((error) => _onError(liveTest, error.error, error.stackTrace)));
+
+    _subscriptions.add(liveTest.onPrint.listen((line) {}));
+  }
+
+  /// Return an ordered list of groups that [liveTest] belongs to
+  List _orderedGroupList(List<Group> listOfGroups) {
+    List separatedList = [];
+    for (int i = listOfGroups.length - 1; i >= 1; i--) {
+      if (i > 1) {
+        separatedList.add(listOfGroups[i]
+            .name
+            .replaceAll(listOfGroups[i - 1].name, '')
+            .trim());
+      } else {
+        separatedList.add((listOfGroups[1].name));
+      }
+    }
+    return (separatedList.reversed).toList();
```

Comment:
instead of building the list backwards and then reversing it at the end, why not do something like this?
```
// probably need some sort of check to verify that the incoming listOfGroups.length > 1
List separatedList = [listOfGroups[1].name];
for (int i = 2; i < listOfGroups.length; i++) {
    separatedList.add(listOfGroups[i]
        .name
        .replaceAll(listOfGroups[i - 1].name, '')
        .trim());
}
return separatedList;
```